src/storage/mod.rs

```diff
@@ -1329,17 +1329,23 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
             callback
         );
 
-        if !self.enable_ttl && ttl != 0 {
-            return Err(Error::from(ErrorInner::TTLNotEnabled));
+        if !self.enable_ttl {
+            for ttl in &ttls {
+                if *ttl != 0 {
+                    return Err(Error::from(ErrorInner::TTLNotEnabled));
+                }
+            }
         }
-        let expire_ts = convert_to_expire_ts(ttl);
+        let mut idx = 0;
 
         let modifies = pairs
             .into_iter()
             .map(|(k, v)| Modify::Put(cf, Key::from_encoded(k), v))
             .map(|mut m| {
                 if self.enable_ttl {
-                    m.with_ttl(expire_ts)
+                    let expire_ts = convert_to_expire_ts(ttls[idx]);
+                    idx += 1;
+                    m.with_ttl(expire_ts);
```

Comment:
Please use [`Iterator::zip()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.zip). And also please check the length of `ttls` and return an error on length mismatch. Or else TiKV may panic.